README.md
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

# sg Node Logger 

## Future Plans (stack ranked)

- Add unit tests
- Data Dog transport
- Explore morgan integration
- Automatically assign SessionId and RequestId if available
- Handling of PII/sensitive information
  - Option 1: Wrap this with fast-redact (https://github.com/davidmarkclements/fast-redact) such that clients can pass keys that should be redacted
  - Option 2: PII tagging system (field/object) to automatically remove sensitive information from logs
- Config files that can be used by consumers to automatically configure logger to fit needs (i.e. using this logger in a raspberry pi env vs in ECS)
- File/env-var based logger on/off switch
  - Useful when logs are not needed (ie, unit tests)
- AWS CloudWatch - auto manage rate-limiting (this may go away with the introduction of the DataDog transport)

## Vision

- Speed up the development of applications without worrying about the logging frameworks
- Make it dead simple to use
- Use uniform log objects
  - No more guessing what the logs look like
  - Simplifies log querying across the board (in Data Dog, for example)

## Warning

> We cannot stress this enough. Do NOT log PII or any sensitive information.
> For any questions, contact our friendly cyber-security team.

## Getting Started

### Log Shape

```ts
{
  // Provided by application/user
  // - changes from log to log
  logLevel: LogLevel;
  message: string;
  tags?: string[];
  data?: any;
  errorMessage?: string;
  errorStack?: string;

  // Dynamic Metadata
  // - susceptible to change from log to log
  utcTimestamp?: string;
  sessionId?: string;
  correlationId?: string;
  requestId?: string;

  // Static Metadata
  // - values are set once during startup
  // - does not change from log to log
  appName?: string;
  appVersion?: string;
  environment?: string;
  hostname?: string;
  ipaddress?: string;
  os?: string;
  osVersion?: string;
  osBuildNumber?: string;
}
```

### Prerequisites

- Add github package registry to your project

```sh
# Because this package is hosted in @sweetgreen's github repository, you will need to add our registry to your project:

# In your project root:
echo "//npm.pkg.github.com/:_authToken=YOUR_GITHUB_PERSONAL_ACCESS_TOKEN\nregistry=https://npm.pkg.github.com/sweetgreen" >> .npmrc

# See https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token to create a personal access token.

# WARNING: make sure the .npmrc file is .gitignored!
```

### Installation

1. Add to your node project

```sh
# yarn
yarn add @sweetgreen/sg-node-logger

# npm
npm i @sweetgreen/sg-node-logger
```

### Environment Variables

The following environment variables are used by the package:

- `NODE_ENV`
  - ONLY when specifying a custom configuration
  - Assigns logger transports based on the environment value

`NODE_ENV` is ignored when using the library out of the box. Meaning, when a custom configuration is not passed to the initializer (`initLogger()`).

## Usage

```javascript
// CommonJS
const { logDebug, initLogger } = require('@sweetgreen/sg-node-logger');

// ES6
import { logDebug, initLogger } from '@sweetgreen/sg-node-logger';

// Initialize logger once and only once
// NOTE: Subsequent usage of `logDebug` does NOT need to be paired with `initLogger()`.
// NOTE: This simple out of the box usage style does not require the presence of `NODE_ENV`.

initLogger('application-name');

// Log away
logDebug('Debug message');
```

#### 🚨 NOTE: When initializing in an application using `dd-trace,` the `dd-trace` initialization should come prior to `initLogger`. See [here](https://sweetgreen.atlassian.net/wiki/spaces/DevOps/pages/1929413719/Advanced+Datadog+Features+Helpful+Tips#Missing-Endpoints-/-URL-Routes) for more information.

### Helper Functions (recommended)

#### Simple Case

```js
import {
  initLogger,
  logDebug,
  logVerbose,
  logInfo,
  logWarn,
  logError,
} from '@sweetgreen/sg-node-logger';

initLogger('application-name');

logDebug('Debug message');
logVerbose('Verbose message');
logInfo('Info message');
logWarn('Warn message');
logError('Error message');
```

#### Logging Custom Data

```js
import { initLogger, logDebug } from '@sweetgreen/sg-node-logger';

initLogger('application-name');

// WARNING: it's your responsibility to remove all PII
const customData = {
  a: 'some string',
  b: 12345,
};

logDebug('Debug message', customData);
```

#### Tagging (highly recommended)

```js
import { initLogger, logDebug } from '@sweetgreen/sg-node-logger';

initLogger('application-name');

const tags: string[] = ['main-feature-name', 'sub-feature-name'];

logDebug('Debug message', undefined, tags);
```

#### Error

```js
import { initLogger, logError } from '@sweetgreen/sg-node-logger';

initLogger('application-name');

try {
  // CODE ...
} catch (error) {
  logError('Error message', error);
}
```

### Configuration

#### Default Configuration

This code is using the default configuration under the hood

```js
import { initLogger, logDebug } from '@sweetgreen/sg-node-logger';

initLogger('application-name');

logDebug('Debug message');
```

Details of the default configuration (for the latest configuration checkout the function `rawJSONConsoleConfig()` in the `configs.ts` file):

```js
{
  environments: [
    {
      nodeEnvName: Environment.All,
      transports: {
        rawJSONConsole: [
          {
            minimumLogLevel: LogLevel.Info,
          } as RawJSONConsoleTransportConfig,
        ],
      },
    },
  ],
}
```

- This configuration is used across all environments (`Environment.All`)

#### Pre-Defined Configurations

- Pre-defined configurations will be added over time to make it as simple as possible to use and to keep the code clean and clear from file based configuration.

```js
import {
  initLogger,
  aPredefinedConfig,
  logDebug,
} from '@sweetgreen/sg-node-logger';

// Keep the configuration close to the application's entry point
initLogger('application-name', aPredefinedConfig);

logDebug('Debug message');
```

#### Custom Configuration

```js
import {
  LoggerOptions,
  SimpleConsoleTransportConfig,
  ColorizedConsoleTransportConfig,
  AwsCloudWatchTransportConfig,
  Transport,
  LogLevel,
  initLogger,
  logDebug,
} from '@sweetgreen/sg-node-logger';

const loggerOptions: LoggerOptions = {
  {
    environments: [
      {
        nodeEnvironmentName: 'production',
        transports: {
          colorizedConsole: [
            {
              minimumLogLevel: LogLevel.Verbose,
            } as ColorizedConsoleTransportConfig,
          ],
          awsCloudWatch: [
            {
              minimumLogLevel: LogLevel.Warn, // Optional: defaults to Info
            } as AwsCloudWatchTransportConfig,
          ],
        },
      },
      {
        nodeEnvironmentName: 'development',
        transports: {
          simpleConsole: [
            {
              minimumLogLevel: LogLevel.Error, // Optional: defaults to Info
            } as SimpleConsoleTransportConfig,
          ],
        },
      },
    ],
  };
}

// Keep the initializer close to the application's entry point
initLogger('application-name', loggerOptions);

logDebug('Debug message');
```

- `nodeEnvironmentName` can be any value that your application uses in its lifecycle ('prod', 'production', 'dev', 'develop', etc) when using the custom configuration.

## Transports

### Available Transports

```js
import {
  LoggerOptions,
  SimpleConsoleTransportConfig,
  ColorizedConsoleTransportConfig,
  AwsCloudwatchTransportConfig,
} from '@sweetgreen/sg-node-logger';

const loggerOptions: LoggerOptions = {
  {
    environments: [
      {
        nodeEnvironmentName: 'production',
        transports: {
          simpleConsole: [
            {
              // configuration goes here
            } as SimpleConsoleTransportConfig,
          ],
          colorizedConsole: [
            {
              // configuration goes here
            } as ColorizedConsoleTransportConfig,
          ],
          awsCloudWatch: [
            {
              // configuration goes here
            } as AwsCloudwatchTransportConfig,
            {
              // same transports can be repeated
            } as AwsCloudwatchTransportConfig,
          ],
          rawJSONConsole: [
            {
              // configuration goes here
            } as RawJSONConsoleTransportConfig,
          ]
        },
      },
    ],
  };
}
```

### SimpleConsole

```
# Sample logs
info: testing info {"data":{},"timestamp":"2020-11-21T06:24:06.048Z"}
warn: testing warn {"data":{},"timestamp":"2020-11-21T06:24:06.048Z"}
error: testing error {"data":{},"timestamp":"2020-11-21T06:24:06.048Z"}
```

### ColorizedConsole

![ColorizedConsole](/assets/pretty-transport-1.png)

### RawJSONConsole

```
{"environment":"production","appName":"sg-node-logger","tags":["startup"],"level":"warn","message":"Verify PII is removed from all logs! Application security is everyone's responsibility.","timestamp":"2021-01-21T17:41:28.365Z"}
{"environment":"production","appName":"sg-node-logger","data":{"a":"jfkkjflsd","b":137843},"level":"info","message":"testing info","timestamp":"2021-01-21T17:41:28.367Z"}
{"environment":"production","appName":"sg-node-logger","tags":["tag1","tag2"],"level":"warn","message":"testing warn","timestamp":"2021-01-21T17:41:28.367Z"}
{"environment":"production","appName":"sg-node-logger","level":"error","message":"testing error","timestamp":"2021-01-21T17:41:28.367Z"}
```

### AwsCloudwatch

- Credentials
  - There are two options:
    - Option 1: read from `~/.aws/credentials`. The logger automatically picks up the credentials from this file. It is the default behavior when using the simple config.
    - Option 2: access key and secret key can be passed to the transport, though manual configuration is required.
- Best Practices
  - `uploadRateInMilliseconds` (must be between 200 and 60000 or 0.2 secs and 60 seconds)
    - Rate Limiting
      - CW Log groups have quotas and rate limits. Make sure it's well understood when deciding this value
        - Currently, CW log streams are limited to 5 requests per second -> 200 milliseconds.
        - [PutLogEvents](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html)
        - [CloudWatch Service Quotas](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html)
    - Values - recommendations
      - Default value is 10000 - 10 seconds
      - Sample usage, use the value 1000 if there will be 5 instances running your application. Which translates to 5 requests per second.
      - Another strategy is to use multiple log streams to avoid rate-limiting.
  - `retentionInDays` (must be between 15 days and 180 days)
    - AWS accepted values = [1, 3, 5, 7, 14, 30, 60, 90, 120, 150, 180, 365, 400, 545, 731, 1827, 3653]
    - Module minimum is 1
    - Module maximum is 180
    - Default is `150` days

Example full custom configuration - typical,

```ts
import {
  AwsCloudwatchTransportConfig,
  LoggerOptions,
  LogLevel,
  initLogger,
  logDebug,
} from '@sweetgreen/sg-node-logger';

const loggerOptions: LoggerOptions = {
  {
    environments: [
      {
        nodeEnvironmentName: 'production',
        transports: {
          awsCloudWatch: [{
            {
              // Optional: defaults to Info
              minimumLogLevel: LogLevel.Info,

              awsRegion: 'us-east-1',

              logGroupName: '/HelloWorldService/Production',

              // Optional: both access and secret must be passed, otherwise
              // it throws an error.
              // Make sure both are empty/undefined to use the '~/.aws/credentials' file.
              accessKeyId: '<aws-access-key-id>',
              secretAccessKey: '<aws-secret-access-key>',

              // Optional: defaults to 10000 (10 secs)
              uploadRateInMilliseconds: 10000,

              // Optional: defaults to 180 days
              retentionInDays: 30,
            } as AwsCloudwatchTransportConfig,
          },
        ],
      },
    ],
  };
}

// Keep the configuration close to the application's entry point (index.ts)
initLogger('application-name', loggerOptions);

logDebug('Debug message');
```

Example using `~/.aws/credentials` + using default values,

```ts
import {
  AwsCloudwatchTransportConfig,
  LoggerOptions,
  initLogger,
  logDebug,
} from '@sweetgreen/sg-node-logger';

const loggerOptions: LoggerOptions = {
  {
    environments: [
      {
        nodeEnvironmentName: 'production',
        transports: {
          awsCloudWatch: [
            {
              awsRegion: 'us-east-1',
              logGroupName: '/HelloWorldService/Production',
            } as AwsCloudwatchTransportConfig,
          ],
        },
      },
    ],
  };
}

// Keep the configuration close to the application's entry point (index.ts)
initLogger('application-name', loggerOptions);

logDebug('Debug message');
```

With this option, the following defaults will be used:

- `minimumLogLevel` will be default `LogLevel.Info`
- `accessKeyId` and `secretAccessKey` are empty, therefore `~/.aws/credentials` will be used
- `uploadRateInMilliseconds` will use the default `10000` ms - 10 seconds
- `retentionInDays` will use the default `180` days

## Contributing

This repo uses commitizen and is set up in such a way that when you commit, you'll be prompted with a commit wizard. Your inputs will determine what version CI will bump this package to 🍻

## Testing

Example of mocking `sg-node-logger`:

```
//  your.test.ts

const { logError } = require('@sweetgreen/sg-node-logger');

// At the top of your file include jest.mock()

jest.mock('@sweetgreen/sg-node-logger', () => {
  return {
    logError: jest.fn(),
    logInfo: jest.fn(),
  };
});


//  Within unit test

   it('throws a console error when message is sent with delay', () => {
      const message = {
        kitchen_update_at: stringDateToUTC(new Date().toISOString(), 1000000),
      };
      const testQueue = 'myQueue';
      const delaySeconds = 0;

      const messageParams = {
        MessageBody: JSON.stringify(message),
        QueueUrl: testQueue,
        DelaySeconds: delaySeconds,
      };

      sendMessageFunc(messageParams);
      expect(logError).toHaveBeenCalled();  //  Alternative example: you could spy on the implementation to check for message
    });

```

See [Run-the-Pass](https://github.com/sweetgreen/run-the-pass/blob/main/lib/aws/__tests__/sqs.test.js) for full example


CHANGELOG.md
## [1.7.8-alpha.3](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.8-alpha.2...v1.7.8-alpha.3) (2023-01-13)


### Bug Fixes

* **security-updates:** resolves existing security vulnerabilities ([6e45bee](https://github.com/sweetgreen/sg-node-logger/commit/6e45bee228b1ee09024dbe24a7e644d7df08529e))

## [1.7.8-alpha.2](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.8-alpha.1...v1.7.8-alpha.2) (2023-01-13)


### Bug Fixes

* **merge-main:** Merge branch 'main' into alpha ([5f515a7](https://github.com/sweetgreen/sg-node-logger/commit/5f515a74cdf155918a96049797e7ae05791d4d07))

## [1.7.8-alpha.1](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.7...v1.7.8-alpha.1) (2023-01-13)


### Bug Fixes

* **no-op:** testing alpha release ([f1dadc5](https://github.com/sweetgreen/sg-node-logger/commit/f1dadc5228621fa5714f7d54f2d33feaa4169bf5))
* **release-pipe:** Update to semantic-release-action@v3 ([20eddcd](https://github.com/sweetgreen/sg-node-logger/commit/20eddcd10aa3008abe34d7e8bc23b9009b5f7836))

## [1.7.7](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.6...v1.7.7) (2022-12-15)


### Bug Fixes

* **deploy-process:** adds husky hooks to package.json ([334f3a2](https://github.com/sweetgreen/sg-node-logger/commit/334f3a2ea38f33586187eba93ac6c67d58834817))
* **deploy-process:** reverts changes to configs.ts ([eae0c5e](https://github.com/sweetgreen/sg-node-logger/commit/eae0c5e4eb94ca956b87ead86106c46bbe97f43c))
* **deploy-process:** testing readding husky precommit hook ([6553b8a](https://github.com/sweetgreen/sg-node-logger/commit/6553b8a69d9461dc299017c2db41e83e7ec1d466))
* **deploy-process:** testing removal of husky precommit hook for github actions ([ceedf9f](https://github.com/sweetgreen/sg-node-logger/commit/ceedf9f7c0d09841f30d6f0cdf08eaa4863500a1))
* **deploy-process:** updates yarn.lock ([56acbf2](https://github.com/sweetgreen/sg-node-logger/commit/56acbf2f3cf67bd81627ebed6daa72e7dfa77c81))
* **transport-cleanup:** cleanup unused refs and code from transports.ts ([81f33c1](https://github.com/sweetgreen/sg-node-logger/commit/81f33c151800d4069904f39b7e1534aeb3833adb))
* **vulnerabilities:** no-op to trigger an alpha build ([62603e1](https://github.com/sweetgreen/sg-node-logger/commit/62603e11dc3cb0041ac6446a6d2f2657ce303c17))

## [1.7.7-alpha.3](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.7-alpha.2...v1.7.7-alpha.3) (2022-12-15)


### Bug Fixes

* **transport-cleanup:** cleanup unused refs and code from transports.ts ([81f33c1](https://github.com/sweetgreen/sg-node-logger/commit/81f33c151800d4069904f39b7e1534aeb3833adb))

## [1.7.7-alpha.2](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.7-alpha.1...v1.7.7-alpha.2) (2022-12-14)


### Bug Fixes

* **deploy-process:** adds husky hooks to package.json ([334f3a2](https://github.com/sweetgreen/sg-node-logger/commit/334f3a2ea38f33586187eba93ac6c67d58834817))
* **deploy-process:** reverts changes to configs.ts ([eae0c5e](https://github.com/sweetgreen/sg-node-logger/commit/eae0c5e4eb94ca956b87ead86106c46bbe97f43c))
* **deploy-process:** testing readding husky precommit hook ([6553b8a](https://github.com/sweetgreen/sg-node-logger/commit/6553b8a69d9461dc299017c2db41e83e7ec1d466))
* **deploy-process:** updates yarn.lock ([56acbf2](https://github.com/sweetgreen/sg-node-logger/commit/56acbf2f3cf67bd81627ebed6daa72e7dfa77c81))

## [1.7.7-alpha.1](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.6...v1.7.7-alpha.1) (2022-09-19)


### Bug Fixes

* **deploy-process:** testing removal of husky precommit hook for github actions ([ceedf9f](https://github.com/sweetgreen/sg-node-logger/commit/ceedf9f7c0d09841f30d6f0cdf08eaa4863500a1))
* **vulnerabilities:** no-op to trigger an alpha build ([62603e1](https://github.com/sweetgreen/sg-node-logger/commit/62603e11dc3cb0041ac6446a6d2f2657ce303c17))

## [1.7.6](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.5...v1.7.6) (2021-02-24)


### Bug Fixes

* **vulnerabilities:** update semantic-release related packages ([01365e1](https://github.com/sweetgreen/sg-node-logger/commit/01365e19bf28b4155db04248e81a99c34248e32d))

## [1.7.5](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.4...v1.7.5) (2021-02-24)


### Bug Fixes

* **vulnerabilities:** update semantic-release which updates marked to a version without vulnerabilities ([eb343a0](https://github.com/sweetgreen/sg-node-logger/commit/eb343a0543efb3ff8508cb8dc609695b2a9bff97))

## [1.7.4](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.3...v1.7.4) (2021-02-24)


### Bug Fixes

* **gh-actions:** trying to fix release auth error ([a8635b4](https://github.com/sweetgreen/sg-node-logger/commit/a8635b435cee199866df0ae70198d07661795a4d))
* **gh-actions:** trying to fix release auth error - 1 ([15d3e8f](https://github.com/sweetgreen/sg-node-logger/commit/15d3e8fcae42088ce6219bb0fcbecd38e7852dec))
* **gh-actions:** trying to fix release auth error - 2 ([ec97065](https://github.com/sweetgreen/sg-node-logger/commit/ec97065af43cac2473a9a3682c3d93aaa84c3441))
* **gh-actions:** trying to fix release auth error - 3 ([77933f5](https://github.com/sweetgreen/sg-node-logger/commit/77933f55e97401f7f04decb8e6ec2604573afb93))

## [1.7.3](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.2...v1.7.3) (2021-02-23)


### Bug Fixes

* **gh-actions:** revert GH_TOKEN usage ([0edc996](https://github.com/sweetgreen/sg-node-logger/commit/0edc9966d382ceab65766a241b412ec41268b3ea))

## [1.7.2](https://github.com/sweetgreen/sg-node-logger/compare/v1.7.1...v1.7.2) (2021-02-23)


### Bug Fixes

* **gh-actions:** add GH_TOKEN ([8630554](https://github.com/sweetgreen/sg-node-logger/commit/8630554466148596fe731d0e5a049061af4d0235))
* **gh-actions:** use GH_TOKEN ([4572531](https://github.com/sweetgreen/sg-node-logger/commit/45725311e7e85d51b4c9f13a3b8acd0bb3fa59b9))

src/index.ts
// NOTE: Special case to allow 'any' in the helper functions ONLY.
// We want to support the passing of any shape for custom data.
/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types */
import dotenv from 'dotenv';
import winston from 'winston';

import {
  AwsCloudWatchTransportConfig,
  SimpleConsoleTransportConfig,
  ColorizedConsoleTransportConfig,
  RawJSONConsoleTransportConfig,
  LogEntry,
  LoggerOptions,
  LogLevel,
} from './types';
import { newLogger, getTransports } from './helpers';
import {
  simpleConsoleConfig,
  colorizedConsoleConfig,
  rawJSONConsoleConfig,
} from './configs';
import { LoggerError } from './errors';

// Init dotenv
dotenv.config();

let logger: winston.Logger;

/**
 * Initializes the logger. Uses default configuration if none is passed.
 *
 * @param options logger options/config
 */
function initLogger(appName: string, options?: LoggerOptions): void {
  const transports = options?.environments
    ? getTransports(appName, options.environments, true)
    : getTransports(appName, rawJSONConsoleConfig());

  logger = newLogger(appName, transports);
}

/**
 * Base log function
 *
 * Note: it's highly recommended to use the helper functions:
 * logDebug, logVerbose, logInfo, logWarn, and logError
 *
 * @param log log object
 */
function log(log: LogEntry): void {
  // TODO: set DynamicLogMetadata
  // const metadata: DynamicLogMetadata = {
  //   utcTimestamp: new Date().toISOString(),
  //   sessionId: 'uuid',
  //   correlationId: 'uuid',
  // };

  const { logLevel, message, ...otherFields } = log;

  const level: string = LogLevel[logLevel].toLowerCase();

  if (!logger) {
    throw new LoggerError(
      `Ensure the logger has been initialized by calling initLogger() at the app's entry point.`
    );
  }

  logger.log(level, message, otherFields);
}

function logDebug(message: string, customData?: any, tags?: string[]): void {
  log({
    logLevel: LogLevel.Debug,
    message: message,
    tags: tags,
    data: customData,
  });
}

function logVerbose(message: string, customData?: any, tags?: string[]): void {
  log({
    logLevel: LogLevel.Verbose,
    message: message,
    tags: tags,
    data: customData,
  });
}

function logInfo(message: string, customData?: any, tags?: string[]): void {
  log({
    logLevel: LogLevel.Info,
    message: message,
    tags: tags,
    data: customData,
  });
}

function logWarn(message: string, customData?: any, tags?: string[]): void {
  log({
    logLevel: LogLevel.Warn,
    message: message,
    tags: tags,
    data: customData,
  });
}

function logError(
  message: string,
  error?: Error,
  customData?: any,
  tags?: string[]
): void {
  log({
    logLevel: LogLevel.Error,
    message: message,
    tags: tags,
    data: customData,
    errorMessage: error?.message,
    errorStack: error?.stack,
  });
}

// Test Only

// // CASE 0: Default
// initLogger('sg-node-logger');

// // CASE 1: Simple Console
// const loggerOptions = {
//   environments: [
//     {
//       nodeEnvironmentName: 'production',
//       transports: {
//         simpleConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as SimpleConsoleTransportConfig,
//         ],
//       },
//     },
//     {
//       nodeEnvironmentName: 'development',
//       transports: {
//         simpleConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as SimpleConsoleTransportConfig,
//         ],
//       },
//     },
//   ],
// };
// initLogger('sg-node-logger', loggerOptions);

// // CASE 2: Colorized Console
// const loggerOptions = {
//   environments: [
//     {
//       nodeEnvironmentName: 'production',
//       transports: {
//         colorizedConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as ColorizedConsoleTransportConfig,
//         ],
//       },
//     },
//     {
//       nodeEnvironmentName: 'development',
//       transports: {
//         colorizedConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as ColorizedConsoleTransportConfig,
//         ],
//       },
//     },
//   ],
// };
// initLogger('sg-node-logger', loggerOptions);

// // CASE 3: AWS - use credentials
// const loggerOptions = {
//   environments: [
//     {
//       nodeEnvironmentName: 'production',
//       transports: [
//         {
//           minimumLogLevel: LogLevel.Info,
//           awsRegion: 'us-east-1',
//           logGroupName: '/Sg-Node-Logger/Dev',
//           uploadRateInMilliseconds: 1000,
//           retentionInDays: 14,
//         } as AwsCloudWatchTransportConfig,
//       ],
//     },
//     {
//       nodeEnvironmentName: 'development',
//       transports: [
//         {
//           minimumLogLevel: LogLevel.Debug,
//         } as ColorizedConsoleTransportConfig,
//       ],
//     },
//   ],
// };
// initLogger('sg-node-logger', loggerOptions);

// // CASE 4: AWS - pass credentials
// const loggerOptions = {
//   environments: [
//     {
//       nodeEnvironmentName: 'production',
//       transports: {
//         awsCloudWatch: [
//           {
//             minimumLogLevel: LogLevel.Info,
//             awsRegion: 'us-east-1',
//             logGroupName: '/Sg-Node-Logger/Dev',
//             accessKeyId: '<aws-access-key-id-goes-here>',
//             secretAccessKey: '<aws-secret-access-key-goes-here>',
//             uploadRateInMilliseconds: 1000,
//             retentionInDays: 1,
//           } as AwsCloudWatchTransportConfig,
//         ],
//       },
//     },
//     {
//       nodeEnvironmentName: 'development',
//       transports: {
//         colorizedConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as ColorizedConsoleTransportConfig,
//         ],
//       },
//     },
//   ],
// };
// initLogger('sg-node-logger', loggerOptions);

// // CASE 5: Run all transports
// const loggerOptions = {
//   environments: [
//     {
//       nodeEnvironmentName: 'development',
//       transports: {
//         simpleConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as SimpleConsoleTransportConfig,
//         ],
//         colorizedConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as ColorizedConsoleTransportConfig,
//         ],
//         awsCloudWatch: [
//           {
//             minimumLogLevel: LogLevel.Info,
//             awsRegion: 'us-east-1',
//             logGroupName: '/Sg-Node-Logger/Dev',
//             uploadRateInMilliseconds: 1000,
//             retentionInDays: 14,
//           } as AwsCloudWatchTransportConfig,
//         ],
//       },
//     },
//   ],
// };
// initLogger('sg-node-logger', loggerOptions);

// // CASE 6: Raw Json
// const loggerOptions = {
//   environments: [
//     {
//       nodeEnvironmentName: 'production',
//       transports: {
//         rawJSONConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as RawJSONConsoleTransportConfig,
//         ],
//       },
//     },
//     {
//       nodeEnvironmentName: 'development',
//       transports: {
//         rawJSONConsole: [
//           {
//             minimumLogLevel: LogLevel.Info,
//           } as RawJSONConsoleTransportConfig,
//         ],
//       },
//     },
//   ],
// };
// initLogger('sg-node-logger', loggerOptions);

// logDebug('testing debug');
// logVerbose('testing verbose');
// logInfo('testing info', { a: 'jfkkjflsd', b: 137843 });
// logWarn('testing warn', undefined, ['tag1', 'tag2']);
// logError('testing error');

export {
  // logger
  initLogger,
  // enums
  LogLevel,
  // interfaces
  LoggerOptions,
  SimpleConsoleTransportConfig,
  ColorizedConsoleTransportConfig,
  AwsCloudWatchTransportConfig,
  RawJSONConsoleTransportConfig,
  // pre-defined configs - more coming soon
  colorizedConsoleConfig,
  simpleConsoleConfig,
  rawJSONConsoleConfig,
  // helper functions
  logDebug,
  logVerbose,
  logInfo,
  logWarn,
  logError,
};
